mermaid.mer

```diff
@@ -0,0 +1,23 @@
+flowchart LR
+  classDef head fill:#ffbfcf;
+  classDef installed fill:#34ebba;
+
+  subgraph amq-online
+    subgraph stable channel
+      amq-online.1.4.0-stable(1.4.0)
+      amq-online.1.4.1-stable(1.4.1) --> amq-online.1.4.0-stable(1.4.0)
+      amq-online.1.4.2-stable(1.4.2) --> amq-online.1.4.1-stable(1.4.1)
+      amq-online.1.4.4-stable(1.4.4) --> amq-online.1.4.2-stable(1.4.2)
+      amq-online.1.5.0-stable(1.5.0)
+      amq-online.1.5.1-stable(1.5.1) --> amq-online.1.4.4-stable(1.4.4)
+      amq-online.1.5.1-stable(1.5.1) --> amq-online.1.5.0-stable(1.5.0)
+      amq-online.1.5.2-stable(1.5.2) --> amq-online.1.5.0-stable(1.5.0)
+      amq-online.1.5.2-stable(1.5.2) --> amq-online.1.5.1-stable(1.5.1)
+      amq-online.1.5.3-stable(1.5.3) --> amq-online.1.5.0-stable(1.5.0)
+      amq-online.1.5.3-stable(1.5.3) --> amq-online.1.5.2-stable(1.5.2)
+      amq-online.1.6.1-stable(1.6.1) --> amq-online.1.5.0-stable(1.5.0)
+      amq-online.1.6.1-stable(1.6.1) --> amq-online.1.5.3-stable(1.5.3)
+      amq-online.1.6.2-stable(1.6.2):::head --> amq-online.1.5.0-stable(1.5.0)
+      amq-online.1.6.2-stable(1.6.2):::head --> amq-online.1.6.1-stable(1.6.1)
+    end
+  end
```

Comment:
IMO we do not need to care too much about it be with the same right logic used by OLM to generate the path.
However, the mermaid needs to following the doc; https://olm.operatorframework.io/docs/contribution-guidelines/upgrade-graphs/oor we need to decide to change the definition.

In the current definition, it is the opposite. I mean we will read the graph from left to right where the head channel which is the upper versioning will b in the right such as a sequence:

1 -> 2 -> 3 -> 4 -> 5 --> 6 (head)

See that in this example we have only replaced. So, one channel replaces another. Your graph is generated here is:

<img width="740" alt="Screen Shot 2021-03-12 at 09 57 15" src="https://user-images.githubusercontent.com/7708031/110924663-38c8bc80-831a-11eb-8d3c-746ef3ee993e.png">

when I would expect:

<img width="755" alt="Screen Shot 2021-03-12 at 10 02 34" src="https://user-images.githubusercontent.com/7708031/110924721-4a11c900-831a-11eb-9dc5-c8b5748e0862.png">

Note that what is expected, at least in imo, is much easier to read and understand. (we read a text from left to right so it makes more sense for our mind. However, for who is native in Chinese, for example, then the opposite might be easier :-)   )

Reply:
Thanks for the great detailed comment @camilamacedo86 I'll work on making the output conform to https://olm.operatorframework.io/docs/contribution-guidelines/upgrade-graphs/oor

To clarify on the specific case of the `amq-online` package from the 4.6 Red Hat index are you saying the bundle is not defined properly, or that it is and just not being rendered properly?

I included `testFile` and `testFile_community` in the repo to help with debugging. In the case of `testFile`, that's the result set rows that are output by sqlite3 for the query file `sqlite3.sql`, So if you search `amq-online` in the `testFile` you can see the raw data from the index.

Reply:
Actually, it's not much text, so here it is:
```csv
amq-online|stable|amq-online.1.6.2|0|1.6.2|>=1.5.4 <1.6.0|amq-online.1.6.1
amq-online|stable|amq-online.1.5.0|1|1.5.0||
amq-online|stable|amq-online.1.6.2|1|1.6.2|>=1.5.4 <1.6.0|amq-online.1.5.0
amq-online|stable|amq-online.1.6.1|2|1.6.1|>=1.5.4 <1.6.0|amq-online.1.5.3
amq-online|stable|amq-online.1.5.0|3|1.5.0||
amq-online|stable|amq-online.1.6.1|3|1.6.1|>=1.5.4 <1.6.0|amq-online.1.5.0
amq-online|stable|amq-online.1.5.3|4|1.5.3||amq-online.1.5.2
amq-online|stable|amq-online.1.5.0|5|1.5.0||
amq-online|stable|amq-online.1.5.3|5|1.5.3||amq-online.1.5.0
amq-online|stable|amq-online.1.5.2|6|1.5.2||amq-online.1.5.1
amq-online|stable|amq-online.1.5.0|7|1.5.0||
amq-online|stable|amq-online.1.5.2|7|1.5.2||amq-online.1.5.0
amq-online|stable|amq-online.1.5.1|8|1.5.1||amq-online.1.4.4
amq-online|stable|amq-online.1.5.0|9|1.5.0||
amq-online|stable|amq-online.1.5.1|9|1.5.1||amq-online.1.5.0
amq-online|stable|amq-online.1.4.4|10|1.4.4||amq-online.1.4.2
amq-online|stable|amq-online.1.4.2|11|1.4.2||amq-online.1.4.1
amq-online|stable|amq-online.1.4.1|12|1.4.1||amq-online.1.4.0
amq-online|stable|amq-online.1.4.0|13|1.4.0||
```

Reply:
@camilamacedo86 so with b2e3076b the graph now looks like this for the amq-online package:
<img width="985" alt="image" src="https://user-images.githubusercontent.com/4449199/110979630-d7f9ae00-8332-11eb-9d3b-993c5eec7223.png">

Reply:
so there is the correct direction to the arrows now. But given the data shown in the table, isn't the above the graph we'd expect? For instance, the table has 1.5.1, 1.5.2, 1.5.3, 1.6.1, 1.6.2 all listed as replacements for 1.5.0 and that's what the graph shows -- in addition to the all the other replaces, 1.4.0 -> 1.6.2

Reply:
The CSV has just one field replace. How the bundle operator 1.5.0 can replace 5 other projects? 
According to the graph and the code added as an example, it has only replaced. So, it is not possible.  

However, looking at the table above https://github.com/operator-framework/index-mermaid-graph/pull/1#discussion_r593176987 it has olm.skipRange as well. So, maybe is that. See that the skipRange has another legend

<img width="791" alt="Screen Shot 2021-03-12 at 23 27 57" src="https://user-images.githubusercontent.com/7708031/111009185-d4921100-838a-11eb-9db0-f152530354d0.png">
<img width="808" alt="Screen Shot 2021-03-12 at 23 28 06" src="https://user-images.githubusercontent.com/7708031/111009190-d5c33e00-838a-11eb-8ea3-b3367fcb5de4.png">

:

Reply:
I will try to create this graph manually and I will let you know.

Reply:
@jmccormick2001 & @gallettilance, @camilamacedo86 was pointing out that while this graph is showing all the replaces, it's not showing edges that would be generated by the `skipRange` entries, they would form parallel edges between many of the nodes linking 1.5.0 to the versions that can upgrade it. Do you think the tool should render both, or have a mode (perhaps default) where it renders both?

Reply:
I'd think the graphs should show the skipRanges somehow.

Reply:
ok well I need another example to see this then, because I made the change to see what it would look like for `amq-online` and then I realized that there are skipRange, but they don't include bundles in the channel or even package. Checking the table above again, I notice `>=1.5.4 <1.6.0` includes none of the bundle versions.

Reply:
okay tried the experimental version with `serverless-operator` as an example where both skipRange and replace edges can be made. It really messes with Mermaid's ability to guess the right place to put the edge label it seems:

<img width="1421" alt="image" src="https://user-images.githubusercontent.com/4449199/111210203-58bee100-85a3-11eb-9cf6-8858067e1f3c.png">

Reply:
see all the skipRange annotations stacked up there to the top-left of the channel?  So I'll see if there is a way to force better drawing...

Reply:
IMO:
- You are getting the skip and skipRange and then, creating the full path ok
- So, instead of use replace for all we could check why that path is created and then use the simbol accordingly. 

BTW the example added above of serverless-operator shows not accurate too.

Reply:
You mean the multiple heads shown in the shown channel @camilamacedo86 ?

Reply:
```So, instead of using replace for all we could check why that path is created and then use the simbol accordingly.```

@camilamacedo86 Are you saying decide which type of edge to create?

Reply:
> You mean the multiple heads shown in the shown channel @camilamacedo86 ?

I am not sure if is possible to have more than one head of the channel. The head is the latest / upper version available in the path. To know the operator bundles which are the latest head we can get only the entries which have values in the `operatorbundle` table for both `csv` and `bundle` column. 

> Regards the image/example in: https://github.com/operator-framework/index-mermaid-graph/pull/1#discussion_r594625397

By looking in the graph I understand that operator bundle 1.8.0 replaces 1.7.2 and also has a skipRange which is missing the range info in the graph but it shows that would be >= 1.7.2 < 1.8.0. Is that correct? Is that the configuration of this operator bundle? Simmilar case for 1.9.0, 1.10.0

> @camilamacedo86 Are you saying decide which type of edge to create?

No. Look in https://olm.operatorframework.io/docs/contribution-guidelines/upgrade-graphs/. We have sumbols to represent the operator bundle configuration. So, the following some questions:

- You are getting all operator bundles and creating the path and then using the replace legend. How we will know what path was made because of a skip or skipRange? How do we know what are the range and values used in these configurations for we know what is wrong and fix?

My suggestion was that this idea might works if we are able to instead use replace for all scenarios use the properly legend with the skip and skipRange values to describe why that path was made. So, the graph would in this scenario be like your first example, however, providing the required information.

Reply:
So b4396e9 fixes the multiple heads per channel bug. Yes, one head per channel, for instance, on the `serverless-operator` package graph now.

And sorry, just one more time to restate to make sure I understand the suggestion: 

When it's possible to upgrade between nodes via both replace and skip or skiprange, use the edge style for replace but with the annotation style of skip, skipRange?

If so, perhaps we need to make that explicit in the legend doc?

Reply:
So, actually, I think 7f30f979 which made the handling of `skipRangeReplace` edges show up like `replace` edges fixed the display issue and it's okay to show both (all 3 eventually) type of edges connecting nodes?

<img width="1314" alt="image" src="https://user-images.githubusercontent.com/4449199/112019936-4e9e6480-8b06-11eb-80a0-b12664888471.png">

What do you think @camilamacedo86 ?

Reply:
Why we have to legends from 1.7.2 to 1.8.0? Is not that made by a skipRange? So, why we have the replace as well? Are both configured for the bundle? I mean, is the 1.8.0 with replace= `1.7.2` and the skipRange?

Reply:
Here's the relevant query result set row (for rh 4.6 index db):
| pkgName | chanName | bndlName | bndl-Depth | bndl-Version | bndlSkipRange | skips | replaceBndlName | head | default_chan |
| --- | --- | --- | --- | --- | --- | --- | --- | --- | --- |
|serverless-operator|4.6|serverless-operator.v1.8.0|4|1.8.0|>=1.7.2 <1.8.0|""|serverless-operator.v1.7.2|serverless-operator.v1.12.0|4.6|

and so the graph shows 1.8.0 can upgrade 1.7.2 via `replace` or `skipRange` I think this is showing it correctly and well.

Reply:
I think we can close this comment as solved.